balancer/rls/internal/picker.go

```diff
@@ -0,0 +1,394 @@
+	// release the read-lock and acquire a write-lock.
+	p.lb.cacheMu.RLock()
+
+	// Lookup data cache and pending request map using request path and keys.
+	cacheKey := cacheKey{path: info.FullMethodName, keys: reqKeys.Str}
+	dcEntry := p.lb.dataCache.getEntry(cacheKey)
+	pendingEntry := p.lb.pendingMap[cacheKey]
+	now := time.Now()
+
+	switch {
+	// No data cache entry. No pending request.
+	case dcEntry == nil && pendingEntry == nil:
+		p.lb.cacheMu.RUnlock()
+		bs := &backoffState{bs: defaultBackoffStrategy}
+		return p.sendRequestAndReturnPick(cacheKey, bs, reqKeys.Map, info)
+
+	// No data cache entry. Pending request exits.
+	case dcEntry == nil && pendingEntry != nil:
+		p.lb.cacheMu.RUnlock()
+		return balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+
+	// Data cache hit. No pending request.
+	case dcEntry != nil && pendingEntry == nil:
+		if dcEntry.expiryTime.After(now) {
+			if !dcEntry.staleTime.IsZero() && dcEntry.staleTime.Before(now) && dcEntry.backoffTime.Before(now) {
+				// Executing the proactive cache refresh in a goroutine simplifies
+				// acquiring and releasing of locks.
+				go func(bs *backoffState) {
+					p.lb.cacheMu.Lock()
+					// It is OK to ignore the return value which indicates if this request
+					// was throttled. This is an attempt to proactively refresh the cache,
+					// and it is OK for it to fail.
+					p.sendRouteLookupRequest(cacheKey, bs, reqKeys.Map, rlspb.RouteLookupRequest_REASON_STALE, dcEntry.headerData)
+					p.lb.cacheMu.Unlock()
+				}(dcEntry.backoffState)
+			}
+			// Delegate to child policies.
+			return p.delegateToChildPolicies(dcEntry, info, func() { p.lb.cacheMu.RUnlock() })
+		}
+
+		// We get here only if the data cache entry has expired. If entry is in
+		// backoff, delegate to default target or fail the pick.
+		if dcEntry.backoffState != nil && dcEntry.backoffTime.After(now) {
+			status := dcEntry.status
+			p.lb.cacheMu.RUnlock()
+			return p.useDefaultPickIfPossible(info, status)
+		}
+
+		// We get here only if the entry has expired and is not in backoff.
+		bs := *dcEntry.backoffState
+		p.lb.cacheMu.RUnlock()
+		return p.sendRequestAndReturnPick(cacheKey, &bs, reqKeys.Map, info)
+
+	// Data cache hit. Pending request exists.
+	default:
+		if dcEntry.expiryTime.After(now) {
+			return p.delegateToChildPolicies(dcEntry, info, func() { p.lb.cacheMu.RUnlock() })
+		}
+		// Data cache entry has expired and pending request exists. Queue pick.
+		p.lb.cacheMu.RUnlock()
+		return balancer.PickResult{}, balancer.ErrNoSubConnAvailable
+	}
+}
+
+// delegateToChildPolicies is a helper function which iterates through the list
+// of child policy wrappers in a cache entry and attempts to find a child policy
+// to which this RPC can be routed to. If there is no child policy in READY
+// state, we delegate to the first child policy arbitrarily.
+//
+// The passed in unlock function is invoked through a defer statement. Some code
```

Comment:
Maybe don't pass an `unlock` function at all and do the `defer` in the caller instead (all current callers always return when calling it)?  Seems like a cleaner design to me that way.

Reply:
Done.